flight/Modules/CharacterOSD/panel.c

```diff
@@ -0,0 +1,471 @@
+{
+	char buffer [PANEL_HORIZON_HEIGHT][PANEL_HORIZON_WIDTH + 1];
+
+	for (uint8_t i = 0; i < PANEL_HORIZON_HEIGHT; i ++) {
+		memcpy(buffer[i], i == PANEL_HORIZON_HEIGHT / 2 ? _horz_center : _horz_line, PANEL_HORIZON_WIDTH);
+		buffer[i][PANEL_HORIZON_WIDTH] = 0;
+	}
+
+	// code below from minoposd
+	int16_t pitch_line = tan(deg2rad(-state->telemetry.attitude_actual.Pitch)) * _PAN_HORZ_LINES;
+	float roll = tan(deg2rad(state->telemetry.attitude_actual.Roll));
+	for (uint8_t col = 1; col <= _PAN_HORZ_INT_WIDTH; col ++) {
+		// center X point at middle of each column
+		float middle = col * _PAN_HORZ_INT_WIDTH - (_PAN_HORZ_INT_WIDTH * _PAN_HORZ_INT_WIDTH / 2) - _PAN_HORZ_INT_WIDTH / 2;
+		// calculating hit point on Y plus offset to eliminate negative values
+		int16_t hit = roll * middle + pitch_line + _PAN_HORZ_LINES;
+		if (hit > 0 && hit < _PAN_HORZ_TOTAL_LINES) {
+			int8_t row = PANEL_HORIZON_HEIGHT - ((hit - 1) / _PAN_HORZ_CHAR_LINES);
+			int8_t subval = (hit - (_PAN_HORZ_TOTAL_LINES - row * _PAN_HORZ_CHAR_LINES + 1)) * _PAN_HORZ_VRES / _PAN_HORZ_CHAR_LINES;
+			if (subval == _PAN_HORZ_VRES - 1)
+				buffer [row - 2][col] = _PAN_HORZ_TOP;
+			buffer[row - 1][col] = _PAN_HORZ_LINE + subval;
+		}
+	}
+
+	for (uint8_t i = 0; i < PANEL_HORIZON_HEIGHT; i ++) {
+		PIOS_MAX7456_puts(state->dev, x, y + i, buffer[i], 0);
+	}
+}
+
+/* Throttle */
+STD_PANEL(THROTTLE, 7, "\x87%d%%", (int)MAX(0, state->telemetry.manual.throttle*100));
+
+/* GroundSpeed */
+STD_PANEL(GROUNDSPEED, 7, "\x80%d\x81", (int)round(pythag(state->telemetry.velocity_actual.North,
+							  state->telemetry.velocity_actual.East)
+						   * 3.6));
+
+STD_PANEL(BATTERYVOLT, 8, "%.2f\x8e", (double)state->telemetry.battery.Voltage);
+
+/* BatCurrent */
+STD_PANEL(BATTERYCURRENT, 8, "%.2f\x8f", (double)state->telemetry.battery.Current);
+
+/* BatConsumed */
+STD_PANEL(BATTERYCONSUMED, 8, "%u\x82", (uint16_t) state->telemetry.battery.ConsumedEnergy);
+
+/* RSSIFlag */
+static void RSSIFLAG_update (charosd_state_t state, uint8_t x, uint8_t y)
+{
+	if (state->telemetry.manual.rssi < 50) {
+		PIOS_MAX7456_put (state->dev, x, y, 0xb4, MAX7456_ATTR_BLINK);
+	}
+}
+
+static void HOMEDISTANCE_update(charosd_state_t state, uint8_t x, uint8_t y)
+{
+	char buffer[8];
+	float dist = pythag(state->telemetry.position_actual.North,
+			    state->telemetry.position_actual.East);
+	if (dist > 1000) {
+		snprintf(buffer, sizeof(buffer), "%.2f%c",
+			 (double)dist/1000, CHAROSD_CHAR_KM);
+	} else {
+		snprintf(buffer, sizeof(buffer), "%d%c",
+			 (int)round(dist), CHAROSD_CHAR_M);
+	}
+
+	terminate_buffer();
+	PIOS_MAX7456_puts(state->dev, x, y, buffer, 0);
+}
+
+#define _ARROWS CHAROSD_CHAR_ARROW_NL
+
+static void HOMEDIRECTION_update(charosd_state_t state, uint8_t x, uint8_t y)
+{
+	float home_dir = (atan2f(state->telemetry.position_actual.East,
+				 state->telemetry.position_actual.North) * RAD2DEG) + 180;
+
+	if (HAS_SENSOR(state->available, HAS_COMPASS)) {
```

Comment:
should we do this, or should we suppress the thing entirely?  If there's no compass, it's suddenly pointing in compass-rose terms.